src/rdkafka_topic.c

```diff
@@ -552,15 +580,71 @@ static int rd_kafka_toppar_leader_update2 (rd_kafka_itopic_t *rkt,
         rktp = rd_kafka_toppar_s2i(s_rktp);
 
         rd_kafka_toppar_lock(rktp);
-        r = rd_kafka_toppar_leader_update(rktp, leader_id, rkb);
+
+        if (leader != NULL &&
+            rktp->rktp_broker != leader &&
+            rktp->rktp_leader_id == leader_id) {
+                rd_kafka_dbg(rktp->rktp_rkt->rkt_rk, TOPIC, "BROKER",
+                        "Topic %s [%"PRId32"]: Leader %"PRId32" unchanged, "
+                        "not migrating away from preferred replica %"PRId32,
+                        rktp->rktp_rkt->rkt_topic->str,
+                        rktp->rktp_partition,
+                        leader_id, rktp->rktp_broker_id);
+                r = 0;
+        } else
+                r = rd_kafka_toppar_broker_update(rktp, leader_id, leader, rd_true);
+
         rd_kafka_toppar_unlock(rktp);
 
-	rd_kafka_toppar_destroy(s_rktp); /* from get() */
+        rd_kafka_toppar_destroy(s_rktp);  /* from get() */
 
 	return r;
 }
 
 
+/**
+ * @brief Revert the topic+partition delegation to the leader.
+ * @returns 1 if the broker delegation was changed, -1 if the broker
+ *          delegation was changed and is now undelegated, else 0.
+ *
+ * @locks caller must have rd_kafka_toppar_lock(rktp)
+ *        and NOT rd_kafka_topic_wrlock(rkt)
+ * @locality any
+ */
+int rd_kafka_toppar_revert_to_leader (rd_kafka_toppar_t *rktp) {
+        rd_kafka_broker_t *leader;
+
+        rd_kafka_topic_wrlock(rktp->rktp_rkt);
+
+        if (rktp->rktp_leader_id == rktp->rktp_broker_id) {
+                rd_kafka_log(rktp->rktp_rkt->rkt_rk, LOG_WARNING, "BROKER",
+                        "Topic %s [%"PRId32"]: Cannot revert toppar to "
```

Comment:
I think it's valid (also imo better) to have this utility function have leader_id == broker_id as a valid precondition - it may be useful in another context in the future where that is an expected input. as is, it happens to be the case that this is never true. I don't see much value in using assert here instead, but i'll change it to decrease the number of things we need to argue about by one.

if seen from my perspective though, this shouldn't be a warning message, it would better be at a debug level (so I did get it wrong whichever way you look at it).